ext/geometry/lib.rs

```diff
@@ -0,0 +1,1532 @@
+  }
+
+  #[fast]
+  #[getter]
+  pub fn right(&self) -> f64 {
+    let x = self.x.get();
+    let width = self.width.get();
+    maximum(x, x + width)
+  }
+
+  #[fast]
+  #[getter]
+  pub fn bottom(&self) -> f64 {
+    let y = self.y.get();
+    let height = self.height.get();
+    maximum(y, y + height)
+  }
+
+  #[fast]
+  #[getter]
+  pub fn left(&self) -> f64 {
+    let x = self.x.get();
+    let width = self.width.get();
+    minimum(x, x + width)
+  }
+}
+
+#[derive(WebIDL)]
+#[webidl(dictionary)]
+pub struct DOMQuadInit {
+  p1: DOMPointInit,
+  p2: DOMPointInit,
+  p3: DOMPointInit,
+  p4: DOMPointInit,
+}
+
+#[derive(Debug)]
+pub struct DOMQuadInner {
+  p1: UnsafeCell<DOMPointInner>,
+  p2: UnsafeCell<DOMPointInner>,
+  p3: UnsafeCell<DOMPointInner>,
+  p4: UnsafeCell<DOMPointInner>,
+}
+
+impl GarbageCollected for DOMQuadInner {}
+
+#[op2]
+impl DOMQuadInner {
+  #[constructor]
+  #[reentrant]
+  #[cppgc]
+  pub fn constructor(
+    #[webidl] p1: DOMPointInit,
+    #[webidl] p2: DOMPointInit,
+    #[webidl] p3: DOMPointInit,
+    #[webidl] p4: DOMPointInit,
+  ) -> DOMQuadInner {
+    #[inline]
+    fn from_point(point: DOMPointInit) -> DOMPointInner {
+      DOMPointInner {
+        inner: RefCell::new(Vector4::new(
+          *point.x, *point.y, *point.z, *point.w,
+        )),
+      }
+    }
+
+    DOMQuadInner {
+      p1: UnsafeCell::new(from_point(p1)),
+      p2: UnsafeCell::new(from_point(p2)),
+      p3: UnsafeCell::new(from_point(p3)),
+      p4: UnsafeCell::new(from_point(p4)),
+    }
+  }
+
+  #[reentrant]
+  #[static_method]
+  #[cppgc]
+  pub fn from_rect(#[webidl] rect: DOMRectInit) -> DOMQuadInner {
+    let DOMRectInit {
+      x,
+      y,
+      width,
+      height,
+    } = rect;
+    DOMQuadInner {
+      p1: UnsafeCell::new(DOMPointInner {
+        inner: RefCell::new(Vector4::new(*x, *y, 0.0, 1.0)),
+      }),
+      p2: UnsafeCell::new(DOMPointInner {
+        inner: RefCell::new(Vector4::new(*x + *width, *y, 0.0, 1.0)),
+      }),
+      p3: UnsafeCell::new(DOMPointInner {
+        inner: RefCell::new(Vector4::new(*x + *width, *y + *height, 0.0, 1.0)),
+      }),
+      p4: UnsafeCell::new(DOMPointInner {
+        inner: RefCell::new(Vector4::new(*x, *y + *height, 0.0, 1.0)),
+      }),
+    }
+  }
+
+  #[reentrant]
+  #[static_method]
+  #[cppgc]
+  pub fn from_quad(#[webidl] quad: DOMQuadInit) -> DOMQuadInner {
+    #[inline]
+    fn from_point(point: DOMPointInit) -> DOMPointInner {
+      DOMPointInner {
+        inner: RefCell::new(Vector4::new(
+          *point.x, *point.y, *point.z, *point.w,
+        )),
+      }
+    }
+
+    DOMQuadInner {
+      p1: UnsafeCell::new(from_point(quad.p1)),
+      p2: UnsafeCell::new(from_point(quad.p2)),
+      p3: UnsafeCell::new(from_point(quad.p3)),
+      p4: UnsafeCell::new(from_point(quad.p4)),
+    }
+  }
+
+  #[cppgc]
+  #[getter]
+  pub fn p1(&self) -> DOMPointInner {
+    // SAFETY: ptr is alive
+    unsafe { ptr::read(self.p1.get()) }
+  }
```

Comment:
@crowlKats I want to use `UnsafeCell` to return the same `DOMPointInner` that `DOMQuadInner` has, but it is cloned. What should I do?

```js
const quad = new DOMQuad({ x: 1 });
console.log(quad.getBounds().width); // 1

quad.p1.x = 2;
console.log(quad.getBounds().width); // expected: 2, actual: 1
```